### PR TITLE
Reimplement a few helpers with new RuntimeHelpers APIs

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Reimplement `InternalBoxEnum` and `AllocateValueType` with the new RuntimeHelpers APIs (SizeOf and Box). This removes some HelperMethodFrames on CoreCLR and shares more code across the runtimes.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Enum.cs [hunk 1]
```diff
@@ -2225,31 +2226,47 @@ public static object ToObject(Type enumType, object value)
...
         public static object ToObject(Type enumType, ulong value) =>
-            InternalBoxEnum(ValidateRuntimeType(enumType), unchecked((long)value));
+            InternalBoxEnum(ValidateRuntimeType(enumType).TypeHandle, unchecked((long)value));
+
+        private static object InternalBoxEnum(RuntimeTypeHandle type, long value)
+        {
+            ReadOnlySpan<byte> rawData = MemoryMarshal.AsBytes(new ReadOnlySpan<long>(ref value));
+            // On little-endian systems, we can always use the pointer to the start of the scratch space
+            // as memory layout since the least-significant bit is at the lowest address.
+            // For bit-endian systems, the least-significant bit is at the highest address, so we need to adjust
```
Comment (filipnavara) [inline src/libraries/System.Private.CoreLib/src/System/Enum.cs:2260]:
```suggestion
            // For big-endian systems, the least-significant bit is at the highest address, so we need to adjust
```

File: src/mono/mono/metadata/icall-def.h [hunk 2]
```diff
@@ -490,8 +489,7 @@ HANDLES(MHAN_1, "GetFunctionPointer", ves_icall_RuntimeMethodHandle_GetFunctionP
 HANDLES(MAHN_3, "ReboxFromNullable", ves_icall_RuntimeMethodHandle_ReboxFromNullable, void, 2, (MonoObject, MonoObjectHandleOnStack))
 HANDLES(MAHN_2, "ReboxToNullable", ves_icall_RuntimeMethodHandle_ReboxToNullable, void, 3, (MonoObject, MonoQCallTypeHandle, MonoObjectHandleOnStack))
 
-ICALL_TYPE(RT, "System.RuntimeType", RT_31)
-HANDLES(RT_31, "AllocateValueType", ves_icall_System_RuntimeType_AllocateValueType, void, 3, (MonoQCallTypeHandle, MonoObject, MonoObjectHandleOnStack))
+ICALL_TYPE(RT, "System.RuntimeType", RT_30)
```
Comment (lambdageek) [inline src/mono/mono/metadata/icall-def.h:492]:
```suggestion
ICALL_TYPE(RT, "System.RuntimeType", RT_1)
```

the identifier is the _first_ icall defined for the class.

File: src/libraries/System.Private.CoreLib/src/System/Enum.cs [hunk 3]
```diff
@@ -2225,31 +2226,47 @@ public static object ToObject(Type enumType, object value)
...
+            // our starting ref to the correct offset from the end of the scratch space to get the value to box.
+            if (!BitConverter.IsLittleEndian)
+            {
+                int size = RuntimeHelpers.SizeOf(type);
+                rawData = rawData.Slice(sizeof(long) - size);
+            }
+
+            return RuntimeHelpers.Box(ref MemoryMarshal.GetReference(rawData), type)!;
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/libraries/System.Private.CoreLib/src/System/Enum.cs:2268]: I wrote some basic benchmarks for Enum.ToObject and actually saw an improvement.

My benchmarks:

```csharp
    private enum E
    {
        A,
        B,
        C
    }

    [Benchmark]
    public object ToObjectSameSize()
    {
        return Enum.ToObject(typeof(E), 1);
    }

    [Benchmark]
    public object ToObjectLarger()
    {
        return Enum.ToObject(typeof(E), 1L);
    }

    [Benchmark]
    public object ToObjectBoxed()
    {
        return Enum.ToObject(typeof(E), (object)1u);
    }
```

Results:

```

BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.4291/22H2/2022Update)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.100-preview.2.24157.14
  [Host]     : .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2
  Job-BRSDCR : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-ASHIII : .NET 9.0.0 (9.0.24.12805), X64 RyuJIT AVX2


```
| Method           | Runtime  | Mean     | Error    | StdDev   | Ratio | RatioSD |
|----------------- |--------- |---------:|---------:|---------:|------:|--------:|
| ToObjectSameSize | CoreRun | 37.40 ns | 0.809 ns | 1.480 ns |  0.90 |    0.03 |
| ToObjectSameSize | .NET 9.0 | 42.90 ns | 0.170 ns | 0.142 ns |  1.00 |    0.00 |
|                  |          |          |          |          |       |         |
| ToObjectLarger   | CoreRun | 38.20 ns | 0.459 ns | 0.407 ns |  0.89 |    0.01 |
| ToObjectLarger   | .NET 9.0 | 43.12 ns | 0.385 ns | 0.341 ns |  1.00 |    0.00 |
|                  |          |          |          |          |       |         |
| ToObjectBoxed    | CoreRun | 40.88 ns | 0.127 ns | 0.106 ns |  0.80 |    0.01 |
| ToObjectBoxed    | .NET 9.0 | 51.13 ns | 0.874 ns | 0.774 ns |  1.00 |    0.00 |

**Review by jkotas — APPROVED** [review]
LGTM. Thanks!

**Comment by LoopedBard3** [comment]
Potential regression: https://github.com/dotnet/perf-autofiling-issues/issues/32739
@jkoritzinsky

**Comment by matouskozak** [comment]
Possible mono regressions:
Linux/arm64 AOT-llvm: https://github.com/dotnet/perf-autofiling-issues/issues/32800

**Comment by jkotas** [comment]
I have moved the issue to dotnet/runtime: https://github.com/dotnet/runtime/issues/101134

**Pull request merged**